runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -148,19 +146,37 @@ public Token appendCheckpoint() {
      */
     @VisibleForTesting
     public Token appendCheckpoint(Token snapshotTimestamp) {
-        appendCheckpointOnSnapshot(snapshotTimestamp);
-        return snapshotTimestamp;
+        return appendCheckpointOnSnapshot(snapshotTimestamp);
     }
 
     /**
      * Write a checkpoint which reflects the state at snapshot.
      *
      * @param snapshotTimestamp snapshot at which the checkpoint is taken.
      */
-    private void appendCheckpointOnSnapshot(Token snapshotTimestamp) {
+    private Token appendCheckpointOnSnapshot(Token snapshotTimestamp) {
         long start = System.currentTimeMillis();
 
-        rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
+        // If snapshotTimestamp (stream tail) is a negative value
+        // we need to enforce a hole to ensure the start log address
+        // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
+        // Having a negative value will prevent trimming from happening.
+
+        // Note: we could simply not checkpoint empty streams with the introduction
+        // of address maps, but for now, we'll keep it consistent with old backpointer
+        // strategy, which required checkpoints for empty streams to avoid TrimmedExceptions
+        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+            snapshotTimestamp = forceHole();
+        }
+
+        try {
+            rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
+        } catch (TrimmedException te) {
```

Comment:
Do you really need this? The idea of writing a whole is to mark a slot in the address space to prevent the sequencer regressing, but if its trimmed, then we can't regress from a trimmed address, right?

Reply:
We do need it, because consider the case where the stream has not been updated for several checkpoint cycles, therefore the stream tail has not changed, but after a couple of cycles let's say it is trimmed and entirely subsumed by a checkpoint. This read will not serve the purpose of hole fill in that case and will throw a TrimmedException, cause it is no longer present.

Reply:
I misunderstood this comment, thinking it was related to catching the TrimmedException but I believe it was actually related to forcing a hole in these cases. Which I agree with, there is no case in filling a hole as there cannot be regression in this case. Removing this!